importC: can't take address of some compound-literals

## duser reported this on 2022-04-24T22:58:28Z

### Transferred from https://issues.dlang.org/show_bug.cgi?id=23053

### CC List

* Walter Bright (@WalterBright)
* 

### Description

````markdown
void fn()
{ 
	int *a = &(int[1]){0};
	int *b = &(int){0};
}

test.c(3): Error: `[0]` is not an lvalue and cannot be modified
test.c(4): Error: cannot modify constant `0`

array one works if & is removed

int one should be the same as taking the address of a local variable typed "int"
````

#### bugzilla (@WalterBright) commented on 2022-09-23T08:48:51Z



````markdown
clang gives a warning:

test.c:4:7: warning: incompatible pointer types initializing 'int *' with an expression of type 'int (*)[1]' [-Wincompatible-pointer-types]
        int *a = &(int[1]){0};
````

--- a/compiler/src/dmd/expressionsem.d
+++ b/compiler/src/dmd/expressionsem.d
@@ -5271,8 +5271,47 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             error(cle.loc, "cannot convert initializer `%s` to expression", toChars(init));
             return setError();
         }
-        result = e;
-        return;
+
+        // leave these to be handled by their respective functions
+        if (e.isStringExp() || e.isStructLiteralExp() || e.isArrayLiteralExp() || e.isAssocArrayLiteralExp())
+        {
+            result = e;
+            return;
+        }
+
+        if (sc.func)
+        {
+            Expression e0;
+            auto ve = extractSideEffect(sc, "__cl", e0, e, true);
+
+            // export (int a = 5, a) so a doesn't forward ref
+            result = new CommaExp(cle.loc, e0, ve).expressionSemantic(sc);
+            return;
+        }
+        else //global variables
+        {
+            Identifier ident = Identifier.generateId("__cl");
+            auto tmp = new VarDeclaration(cle.loc, t, ident, new ExpInitializer(cle.loc, e));
+
+            // static const type ???
+            tmp.storage_class = STC.static_ | STC.const_ | STC.ctfe;
+
+            // for our global variables, we need them in our symbol table
+            if (sc._module)
+            {
+                sc._module.members.push(tmp);
+            }
+            else if (sc.minst)
+            {
+                sc.minst.members.push(tmp);
+            }
+
+            dsymbolSemantic(tmp, sc);
+
+            auto ve = new VarExp(cle.loc, tmp).expressionSemantic(sc);
+            result = ve;
+            return;
+        }
     }
 
     override void visit(TypeExp exp)
@@ -9181,7 +9220,7 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
              * since C regards string/array literals as lvalues
              */
             auto e = exp.e1;
-            if(e.isStringExp() || e.isArrayLiteralExp())
+            if (e.isStringExp() || e.isArrayLiteralExp())
             {
                 e.type = typeSemantic(e.type, Loc.initial, sc);
                 // if type is already a pointer exp is an illegal expression of the form `&(&"")`

--- a/compiler/test/runnable/fix20092.c
+++ b/compiler/test/runnable/fix20092.c
@@ -1,0 +1,27 @@
+#include <assert.h>
+
+int *m = &(int){0};
+int *p = &(int){8880};
+int *q = &(int){9334};
+
+
+int main()
+{
+    //global symbols
+    assert(*m == 0);
+    assert(*p == 8880);
+    assert(*q == 9334);
+
+    //local symbols
+    int *a = &(int){0};
+    int *b = &(int){55};
+    int *c = &(int){90};
+    assert(*a == 0);
+    assert(*b == 55);
+    assert(*c == 90);
+    *b = 100;
+    *c = 506;
+    assert(*b == 100);
+    assert(*c == 506);
+    return 0;
+}

